### DISCUSSION TITLE
Why do I get "Credential is missing" error when uploading a pdf file to S3 by the js s3 client?

### DISCUSSION THREAD

**Discussion by xinluleo** [aws/aws-sdk-js-v3#4337] [Q&A]
Hi, I am trying to upload a pdf file to my S3 bucket. It should be an easy task. I followed the example here: https://github.com/awsdocs/aws-doc-sdk-examples/blob/main/javascriptv3/example_code/s3/src/s3_upload_object.js, but got the following error:
```console
Error Error: Credential is missing
    at SignatureV4.credentialProvider (runtimeConfig.browser.js:28:1)
    at SignatureV4.signRequest (SignatureV4.js:83:1)
    at SignatureV4.sign (SignatureV4.js:52:1)
    at SignatureV4MultiRegion.sign (SignatureV4MultiRegion.js:13:1)
    at middleware.js:12:1
    at async retryMiddleware.js:24:1
    at async flexibleChecksumsMiddleware.js:53:1
    at async loggerMiddleware.js:2:1
```

My code is attached below:
src/lib/s3client.tsx:
```js
import { S3Client } from '@aws-sdk/client-s3'
// Set the AWS Region.
const REGION = 'us-east-1'
// Create an Amazon S3 service client object.
const s3Client = new S3Client({ region: REGION })
export { s3Client }
```
upload util function:
```js
import { PutObjectCommand } from '@aws-sdk/client-s3'
import { PDFDocument } from 'pdf-lib'

import { s3Client } from 'src/lib/s3client'

export const uploadPDF = async (pdfData, prefix, filename) => {
  const pdfDoc = await PDFDocument.load(pdfData)
  const data = await pdfDoc.saveAsBase64()

  const uploadParams = {
    Bucket: 'fortue-files',
    Key: prefix + '/' + filename,
    Body: data,
    ContentEncoding: 'base64',
    ContentType: 'application/pdf',
  }

  try {
    const response = await s3Client.send(new PutObjectCommand(uploadParams))
    console.log('Success', response)
    return response
  } catch (error) {
    console.log('Error', error)
  }
}
```

The fortue-files bucket is created with all default settings. What is the credential I am missing here?

**Comment by RanVaknin** [comment]
Hi @xinluleo ,

For the SDK to sign your requests, it needs to draw credentials from somewhere. Here is [an article](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-credentials-node.html) of all the different ways you can load credentials.
If this is your first time using the SDK, I suggest you start with [configuring your shared credentials](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/loading-node-credentials-shared.html). This way you don't have to worry about credentials ever again as long as your credentials have not expired.

Let me know if you have any other questions.

Thanks,
Ran

**Comment by AQS-DTheuke** [comment]
For me the issue was, that I used the "bundled" sdk classes instead of aws provided ones.

I had to add the aws-sdk/* to the external rollupOptions.

Vite-Config:

````ts
// https://vite.dev/config/
export default defineConfig({
  build: {
    [...]
    rollupOptions: {
      external: [
        '@aws-sdk/client-s3',
        '@aws-sdk/client-codeartifact',
        '@aws-sdk/client-cloudfront',
        ...builtinModules,
      ],
    },
    [...]
  },
});
````

**Reply by adam-nielsen** [reply]
Thanks this solved my problem!  You can also use regex so you don't have to list every single SDK package you're using, or remember to keep it up to date as you add new ones:
```js
rollupOptions: {
  external: [
    // Don't bundle the AWS SDK, because it's available on Lambda already,
    // and also the bundled version can't get AWS creds properly on Lambda.
    /@aws-sdk\/.*/,
  ],
},
```

**Comment by LiZi52** [comment]
> Hi, I am trying to upload a pdf file to my S3 bucket. It should be an easy task. I followed the example here: https://github.com/awsdocs/aws-doc-sdk-examples/blob/main/javascriptv3/example_code/s3/src/s3_upload_object.js, but got the following error:
> 
> ```console
> 
> Error Error: Credential is missing
> 
>     at SignatureV4.credentialProvider (runtimeConfig.browser.js:28:1)
> 
>     at SignatureV4.signRequest (SignatureV4.js:83:1)
> 
>     at SignatureV4.sign (SignatureV4.js:52:1)
> 
>     at SignatureV4MultiRegion.sign (SignatureV4MultiRegion.js:13:1)
> 
>     at middleware.js:12:1
> 
>     at async retryMiddleware.js:24:1
> 
>     at async flexibleChecksumsMiddleware.js:53:1
> 
>     at async loggerMiddleware.js:2:1
> 
> ```
> 
> 
> 
> My code is attached below:
> 
> src/lib/s3client.tsx:
> 
> ```js
> 
> import { S3Client } from '@aws-sdk/client-s3'
> 
> // Set the AWS Region.
> 
> const REGION = 'us-east-1'
> 
> // Create an Amazon S3 service client object.
> 
> const s3Client = new S3Client({ region: REGION })
> 
> export { s3Client }
> 
> ```
> 
> upload util function:
> 
> ```js
> 
> import { PutObjectCommand } from '@aws-sdk/client-s3'
> 
> import { PDFDocument } from 'pdf-lib'
> 
> 
> 
> import { s3Client } from 'src/lib/s3client'
> 
> 
> 
> export const uploadPDF = async (pdfData, prefix, filename) => {
> 
>   const pdfDoc = await PDFDocument.load(pdfData)
> 
>   const data = await pdfDoc.saveAsBase64()
> 
> 
> 
>   const uploadParams = {
> 
>     Bucket: 'fortue-files',
> 
>     Key: prefix + '/' + filename,
> 
>     Body: data,
> 
>     ContentEncoding: 'base64',
> 
>     ContentType: 'application/pdf',
> 
>   }
> 
> 
> 
>   try {
> 
>     const response = await s3Client.send(new PutObjectCommand(uploadParams))
> 
>     console.log('Success', response)
> 
>     return response
> 
>   } catch (error) {
> 
>     console.log('Error', error)
> 
>   }
> 
> }
> 
> ```
> 
> 
> 
> The fortue-files bucket is created with all default settings. What is the credential I am missing here?

**Reply by LiZi52** [reply]
AWS
